Userland/Libraries/LibGUI/FileSystemModel.cpp

```diff
@@ -668,43 +671,65 @@ bool FileSystemModel::fetch_thumbnail_for(Node const& node)
         return true;
     }
 
-    // Otherwise, arrange to render the thumbnail
-    // in background and make it available later.
+    // See if the thumbnail was loaded, but the callback was never invoked.
+    // This often happens if a window with a FileSystemModel is closed and later reopened.
+    auto promise_it = s_loading_thumbnails.find(path);
+    if (promise_it != s_loading_thumbnails.end()) {
+        auto const& result = promise_it->value->result();
+        // The promise was successfully resolved since we started it.
+        if (result.has_value()) {
+            s_loading_thumbnails.remove(path);
+            s_thumbnail_cache.set(path, result.value());
+            node.thumbnail = s_thumbnail_cache.get(path).value();
+            return true;
+        }
+        // The promise was cancelled, load this thumbnail again.
+        if (promise_it->value->is_cancelled()) {
+            s_loading_thumbnails.remove(path);
+        } else {
+            return false;
+        }
```

Comment:
This is a bit confusingly factored. We have if it was succesfully loaded, if it's still loading, and if it was cancelled.

```
// The promise is still loading, let it complete in the background
if (promise_it->value->is_pending())
    return false;

// The promise was cancelled. Remove it from the map and re-start the load
s_loading_thumbnails.remove(path);
```

But... why don't we just remove it from the loading thumbnails when it's cancelled? instead of leaving it in the map? 

We remove it from s_loading_thumbnails when it's resolved...

When can the above condition happen, where there's a promise on the bg thread that is already resolved, but its resolution callback hasn't gone through? This feels like a race condition that we don't want to mess with.

Reply:
> But... why don't we just remove it from the loading thumbnails when it's cancelled? instead of leaving it in the map? 

I tried this, and it will send the image into an endless reload-fail loop, I'm not quite sure why.

> When can the above condition happen, where there's a promise on the bg thread that is already resolved, but its resolution callback hasn't gone through? This feels like a race condition that we don't want to mess with.

Yep, will fix